src/State/DelayScheduler.lua

```diff
@@ -0,0 +1,70 @@
+--!strict
+
+--[[
+  Manages batch updating of delay objects.
+]]
+
+local RunService = game:GetService("RunService")
+
+local Package = script.Parent.Parent
+local Types = require(Package.Types)
+local updateAll = require(Package.Dependencies.updateAll)
+
+type Set<T> = {[T]: any}
+type Delay = Types.Delay<any>
+type DelayData = {
+  delay: Delay,
+  start: number,
+  duration: number,
+  goal: any
+}
+
+local DelayScheduler = {}
+
+-- all delays to be updated
+local allDelays: Set<DelayData> = {}
+
+--[[
+  Adds a Delay to be updated every render step.
+]]
+function DelayScheduler.add(delay: Delay)
+  allDelays[{
+    delay = delay,
+    start = os.clock(),
+    duration = if delay._durationIsState then delay._duration:get(false) else delay._duration,
+    goal = delay._nextValue
+  }] = true
+end
+
+--[[
+  Removes a Delay from the scheduler.
+]]
+function DelayScheduler.remove(delayData: DelayData)
+  allDelays[delayData] = nil
+end
+
+--[[
+  Updates all Delay objects.
+]]
+local function updateAllDelays()
+  local now = os.clock()
+
+  for delayObject in pairs(allDelays) do
+    local delay = delayObject.delay
+    local currentTime = now - delayObject.start
+    
+    if currentTime >= delayObject.duration then
+      delay._currentValue = delayObject.goal
+      updateAll(delay)
+      DelayScheduler.remove(delayObject)
+    end
+  end
+end
+
+RunService:BindToRenderStep(
+  "__FusionDelayScheduler",
+  Enum.RenderPriority.First.Value,
+  updateAllDelays
+)
+
+return DelayScheduler
```

Comment:
Just thinking about it now, why is DelayScheduler even necessary with this implementation (the way Duration state works)
All that's necessary is ```lua
task.delay(
    if delay._durationIsState then delay._duration:get(false) else delay._duration,
    function()
        delay._currentValue = goal
    end,
)

Reply:
I believe you're right in saying that duration shouldn't be a state object. I implemented the functionality proposed in the original issue without any thought but it causes confusion and more problems than necessary. Also I had never used the `delay()` or `task.delay()` globals as I didn't know how "safe" they were but it's good to know the task library has a good implementation. If we're in agreement I'll remove the DelayScheduler and transition to a Delay object that only accepts a constant duration.

Reply:
I'd hold off on switching to task.delay - It's not used inside Fusion yet as far as I know, and would be a new concept. Elttob may want to consider that first before it's added.

Reply:
> I'd hold off on switching to task.delay - It's not used inside Fusion yet as far as I know, and would be a new concept. Elttob may want to consider that first before it's added.

I missed this reply and just committed 2d555a7 which uses `task.delay()` and removes the DelayScheduler. This can easily be reverted though. @Elttob what are your opinions on `task.delay()`? Should a whole scheduler be used for such a primitive task?

Reply:
If it works reliably and it keeps the code simpler, I don't see a problem with it :)